networkit/cpp/centrality/DynBetweennessOneNode.cpp

```diff
@@ -27,51 +30,55 @@ void DynBetweennessOneNode::run() {
     sigma.resize(G.upperNodeIdBound());
     sigmax.resize(G.upperNodeIdBound());
     Pred.resize(G.upperNodeIdBound());
-    // for each node, we execute a BFS and compute number of SPs between node pairs and number of
-    // SPs between node pairs that go through x
-    G.forNodes([&](node source) {
-        distances[source].resize(G.upperNodeIdBound(), infDist);
-        sigma[source].resize(G.upperNodeIdBound(), 0);
-        sigmax[source].resize(G.upperNodeIdBound(), 0);
-        std::queue<node> q;
-        std::vector<bool> visited(G.upperNodeIdBound(), false);
-        q.push(source);
-        visited[source] = true;
-        sigma[source][source] = 1;
-        distances[source][source] = 0;
-        while (!q.empty()) {
-            node u = q.front();
-            q.pop();
-            if (u == x) {
-                sigmax[source][u] = sigma[source][u];
-            }
-            // insert untouched neighbors into queue
-            G.forNeighborsOf(u, [&](node v) {
-                if (!visited[v]) {
-                    q.push(v);
-                    visited[v] = true;
-                    distances[source][v] = distances[source][u] + 1;
-                }
-                if (distances[source][v] == distances[source][u] + 1) {
-                    // all the shortest paths to u are also shortest paths to v now
-                    sigma[source][v] += sigma[source][u];
-                    if (u == x) {
-                        sigmax[source][v] = sigma[source][u];
-                    } else {
-                        sigmax[source][v] += sigmax[source][u]; // it is 0 until we visit x
-                    }
-                }
-            });
+
+    auto computeDependencies = [&](node s) {
+        // run SSSP algorithm and keep track of everything
+
+        distances[s].resize(G.upperNodeIdBound(), infDist);
+        sigma[s].resize(G.upperNodeIdBound(), 0);
+        sigmax[s].resize(G.upperNodeIdBound(), 0);
+
+        std::unique_ptr<SSSP> sssp;
+        if (G.isWeighted()) {
+            sssp = std::make_unique<Dijkstra>(G, s, true, true);
+        } else {
+            sssp = std::make_unique<BFS>(G, s, true, true);
         }
-    });
 
-    G.forNodes([&](node s) {
+        sssp->run();
+
+        // initialize sigma and distances for source s
+        G.forNodes([&](node t) {
+            distances[s][t] = sssp->distance(t);
+            sssp->numberOfPaths(t).ToDouble(sigma[s][t]);
+        });
+
+        // initialize sigmax (paths that pass through x) for source s
         G.forNodes([&](node t) {
-            if (t != x && s != x && sigma[s][t] != 0) {
-                bcx += double(sigmax[s][t]) / sigma[s][t];
+            auto paths = sssp->getPaths(t);
+            for (auto path : paths) {
+                if (std::find(path.begin(), path.end(), x) != path.end()) {
```

Comment:
Have you measured the time complexity of this change? There can be a lot of paths for dense networks and `find` takes linear time.

Reply:
The `run()` method now produces correct betweenness scores for weighted graphs, but indeed, takes longer than previously. We could practically have both variants implemented and choose the one suited for the input graph (not the prettiest solution, but for the unweighted case runtimes should remain similar). 

On the other hand, our current `SSSP` variants do not store the shortest paths explicitly (rather they only store the predecessor(s) for each node). This makes computing the `bc(x)` complicated without designing a new `BFS-like` algorithm that computes `bc(x)` on-the-fly. Here, any recommendations on your behalf are welcome.

Reply:
The `run()` method has been adjusted to use a priority queue now instead of a regular queue. The runtime takes a slight hit, but the additional support for weighted graphs should outweigh that cost.